[Bug]: AWS Glue Trigger

### Terraform Core Version

1.4.6

### AWS Provider Version

4.66.0_x5

### Affected Resource(s)

aws_glue_trigger

### Expected Behavior

no error.

### Actual Behavior

terraform plan looks good, but apply give panic error

### Relevant Error/Panic Output Snippet

```shell
│ Error: Plugin did not respond
│
│   with module.glue.aws_glue_trigger.initial_trigger["ft1-step_2"],
│   on ../../modules/glue/glue_trigger.tf line 1, in resource "aws_glue_trigger" "initial_trigger":
│    1: resource "aws_glue_trigger" "initial_trigger" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
Releasing state lock. This may take a few moments...

Stack trace from the terraform-provider-aws_v4.66.0_x5 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 383 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/glue.expandActions({0x14003db8af0?, 0x1, 0x0?})
	github.com/hashicorp/terraform-provider-aws/internal/service/glue/trigger.go:477 +0x720
github.com/hashicorp/terraform-provider-aws/internal/service/glue.resourceTriggerCreate({0x11266aae0?, 0x14004924e40}, 0x14003e4ea00, {0x112511600?, 0x140001f4c00})
	github.com/hashicorp/terraform-provider-aws/internal/service/glue/trigger.go:214 +0xfc
github.com/hashicorp/terraform-provider-aws/internal/provider.interceptedHandler[...].func1(0x0?, {0x112511600?, 0x140001f4c00?})
	github.com/hashicorp/terraform-provider-aws/internal/provider/intercept.go:95 +0x118
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x11266aae0?, {0x11266aae0?, 0x140045f3bc0?}, 0xd?, {0x112511600?, 0x140001f4c00?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:702 +0x64
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x140012e1b20, {0x11266aae0, 0x140045f3bc0}, 0x140007e3790, 0x14003e4e880, {0x112511600, 0x140001f4c00})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:837 +0x86c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x14001c13710, {0x11266aae0?, 0x140045f3aa0?}, 0x14003230280)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/grpc_provider.go:1021 +0xb70
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ApplyResourceChange({0x14002c85020, 0x14002c85080, {0x14004f75940, 0x2, 0x2}, {0x0, 0x0, 0x0}, {0x0, 0x0, ...}, ...}, ...)
	github.com/hashicorp/terraform-plugin-mux@v0.10.0/tf5muxserver/mux_server_ApplyResourceChange.go:27 +0xdc
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x1400265ab40, {0x11266aae0?, 0x140045f2870?}, 0x14003e3c1c0)
	github.com/hashicorp/terraform-plugin-go@v0.15.0/tfprotov5/tf5server/server.go:818 +0x3b8
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1121fede0?, 0x1400265ab40}, {0x11266aae0, 0x140045f2870}, 0x14003e3c150, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.15.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:419 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0x1400497e3c0, {0x11267afc0, 0x140007da680}, 0x14003cd67e0, 0x1400507b980, 0x11913a1a0, 0x0)
	google.golang.org/grpc@v1.54.0/server.go:1345 +0xc64
google.golang.org/grpc.(*Server).handleStream(0x1400497e3c0, {0x11267afc0, 0x140007da680}, 0x14003cd67e0, 0x0)
	google.golang.org/grpc@v1.54.0/server.go:1722 +0x82c
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.54.0/server.go:966 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.54.0/server.go:964 +0x290

Error: The terraform-provider-aws_v4.66.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Terraform Configuration Files

```hcl
# autoconf tfvar file:
file_template = {
    ft1 = {
      etl_glue = {
        workflow = {
          job = {
            step_types = {
              step_1 = {
                service   = "job"
                name      = "step_1"
                type      = "ON_DEMAND"
                predicate = {}
                actions = {
                  job_name = "ft1-step_1"
                }
              },
              step_2 = {
                service = "crawler"
                name    = "step_2"
                type    = "CONDITIONAL"
                predicate = {
                  state    = "SUCCEEDED"
                  job_name = "ft1-step_1"
                }
                actions = {
                  crawler_name = "ft1"
                }
              },
            }
          }
          }
        }
      }
    }
  }
```
```hcl
locals {
  job_steps = flatten([
    for job, atr in var.properties.file_template : [
      for job2, atr2 in atr.etl_glue.workflow.job.step_types : {
        name      = job
        service   = atr2.service
        job_step  = atr2.name
        type      = atr2.type
        predicate = atr2.predicate
        actions   = atr2.actions
      }
    ]
  ])
}
```
```hcl
resource "aws_glue_trigger" "initial_trigger" {
  for_each      = { for step in local.job_steps : "${step.name}-${step.job_step}" => step }
  name          = "${var.tags["env"]}-${var.tags["product"]}-${var.tags["service"]}-${each.value.job_step}-job-trigger"
  type          = each.value.type
  description   = "AWS Glue Trigger for ${upper(var.tags["product"])} ETL - ${each.value.job_step}"
  workflow_name = aws_glue_workflow.glue_workflow[var.tags["process"]].name
  enabled       = true

  dynamic "actions" {
    iterator = actions
    for_each = each.value.actions
    content {
      job_name     = actions.key == "job_name" ? aws_glue_job.python_job[each.key].name : null
      crawler_name = actions.key == "crawler_name" ? aws_glue_crawler.target_aurora_db[each.value.name].name : null
    }
  }

  dynamic "predicate" {
    for_each = each.value.type == "ON_DEMAND" ? {} : var.status
    content {
      conditions {
        state    = each.value.predicate.state
        job_name = aws_glue_job.python_job[each.value.predicate.job_name].name
      }
    }
  }
  tags = var.tags
}

``` 

### Steps to Reproduce

remove state lock file and init again

### Debug Output

_No response_

### Panic Output

[tf_log.log](https://github.com/hashicorp/terraform-provider-aws/files/11405521/tf_log.log)


### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

None

--- a/.changelog/38994.txt
+++ b/.changelog/38994.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_glue_trigger: Fix crash when null `action` is configured
+```

--- a/internal/service/glue/trigger.go
+++ b/internal/service/glue/trigger.go
@@ -485,6 +485,9 @@ func expandActions(l []interface{}) []awstypes.Action {
 	actions := []awstypes.Action{}
 
 	for _, mRaw := range l {
+		if mRaw == nil {
+			continue
+		}
 		m := mRaw.(map[string]interface{})
 
 		action := awstypes.Action{}
@@ -520,6 +523,9 @@ func expandActions(l []interface{}) []awstypes.Action {
 }
 
 func expandTriggerNotificationProperty(l []interface{}) *awstypes.NotificationProperty {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
 	m := l[0].(map[string]interface{})
 
 	property := &awstypes.NotificationProperty{}
@@ -535,6 +541,9 @@ func expandConditions(l []interface{}) []awstypes.Condition {
 	conditions := []awstypes.Condition{}
 
 	for _, mRaw := range l {
+		if mRaw == nil {
+			continue
+		}
 		m := mRaw.(map[string]interface{})
 
 		condition := awstypes.Condition{
@@ -564,6 +573,9 @@ func expandConditions(l []interface{}) []awstypes.Condition {
 }
 
 func expandPredicate(l []interface{}) *awstypes.Predicate {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
 	m := l[0].(map[string]interface{})
 
 	predicate := &awstypes.Predicate{
@@ -664,6 +676,9 @@ func flattenTriggerNotificationProperty(property *awstypes.NotificationProperty)
 }
 
 func expandEventBatchingCondition(l []interface{}) *awstypes.EventBatchingCondition {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
 	m := l[0].(map[string]interface{})
 
 	ebc := &awstypes.EventBatchingCondition{

--- a/internal/service/glue/trigger_test.go
+++ b/internal/service/glue/trigger_test.go
@@ -8,6 +8,7 @@ import (
 	"fmt"
 	"testing"
 
+	"github.com/YakDriver/regexache"
 	"github.com/aws/aws-sdk-go-v2/aws"
 	awstypes "github.com/aws/aws-sdk-go-v2/service/glue/types"
 	sdkacctest "github.com/hashicorp/terraform-plugin-testing/helper/acctest"
@@ -462,6 +463,27 @@ func TestAccGlueTrigger_Actions_security(t *testing.T) {
 				ImportState:             true,
 				ImportStateVerify:       true,
 				ImportStateVerifyIgnore: []string{names.AttrEnabled},
+			},
+		},
+	})
+}
+
+// Ensure a null action does not trigger a panic
+// Ref: https://github.com/hashicorp/terraform-provider-aws/issues/31213
+func TestAccGlueTrigger_Actions_null(t *testing.T) {
+	ctx := acctest.Context(t)
+
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.GlueServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckTriggerDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccTriggerConfig_actionsNull(rName),
+				ExpectError: regexache.MustCompile(`InvalidInputException: Actions cannot be null or empty`),
 			},
 		},
 	})
@@ -939,6 +961,17 @@ resource "aws_glue_trigger" "test" {
     batch_size   = 1
     batch_window = 50
   }
+}
+`, rName))
+}
+
+func testAccTriggerConfig_actionsNull(rName string) string {
+	return acctest.ConfigCompose(testAccJobConfig_required(rName), fmt.Sprintf(`
+resource "aws_glue_trigger" "test" {
+  name = %[1]q
+  type = "ON_DEMAND"
+
+  actions {}
 }
 `, rName))
 }

